trin-execution/src/trie_walker/mod.rs

```diff
@@ -0,0 +1,276 @@
+                let extension = extension.read().expect("Extension node must be readable");
+                self.process_node(
+                    extension.node.clone(),
+                    proof.clone(),
+                    [
+                        path.as_slice(),
+                        extension.prefix.get_data().to_vec().as_slice(),
+                    ]
+                    .concat(),
+                )
+                .expect("Failed to process node");
+            }
+            Node::Branch(branch) => {
+                let branch = branch.read().expect("Branch node must be readable");
+
+                // We want to iterate over the children in reverse order so that we can push them to
+                // the stack in order
+                for (i, child) in branch.children.iter().enumerate().rev() {
+                    self.process_node(
+                        child.clone(),
+                        proof.clone(),
+                        [path.as_slice(), &[i as u8]].concat(),
+                    )
+                    .expect("Failed to process node");
+                }
+            }
+            // If the node is a leaf node, we don't need to go deeper
+            _ => {}
+        }
+
+        Some(next_proof)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use eth_trie::{EthTrie, RootWithTrieDiff, Trie};
+    use revm_primitives::{keccak256, Address, B256, U256};
+    use std::{str::FromStr, sync::Arc};
+    use tracing_test::traced_test;
+    use trin_utils::dir::create_temp_test_dir;
+
+    use crate::{
+        config::StateConfig,
+        execution::TrinExecution,
+        storage::{trie_db::TrieRocksDB, utils::setup_rocksdb},
+        utils::full_nibble_path_to_address_hash,
+    };
+
+    #[tokio::test]
+    #[traced_test]
+    async fn test_state_walker() {
+        let temp_directory = create_temp_test_dir().unwrap();
+        let db = Arc::new(setup_rocksdb(temp_directory.path()).unwrap());
+        let mut trie = EthTrie::new(Arc::new(TrieRocksDB::new(false, db.clone())));
+
+        for i in 1..=18 {
+            trie.insert(
+                B256::from(U256::from(i)).as_slice(),
+                B256::from(U256::from(i)).as_slice(),
+            )
+            .unwrap();
+        }
+
+        let root_hash = trie.root_hash().unwrap();
+        let walker = TrieWalker::new(root_hash, trie.db.clone()).unwrap();
+        let mut count = 0;
+        let mut leaf_count = 0;
+        for proof in walker {
+            count += 1;
+
+            let Some(encoded_last_node) = proof.proof.last() else {
+                panic!("Account proof is empty");
+            };
+
+            let Node::Leaf(leaf) =
+                decode_node(&mut encoded_last_node.as_ref()).expect("Failed to decode node")
+            else {
+                continue;
+            };
+            leaf_count += 1;
+
+            // reconstruct the address hash from the path so that we can fetch the
```

Comment:
nit: this comment is still not clear to me. what are we fetching from the database?
aren't we fetching proof directly from the trie?